docs/proposals/pod-security-context.md

```diff
@@ -0,0 +1,302 @@
+container, it requires more complex validations and allows for transcription errors between
+containers
+
+Additionally, use cases for sharing resources such as volumes are much easier to implement if a user
+can describe a pod-level security context.
+
+### Use Case: Override pod security context for container
+
+Some use cases require the containers in a pod to run with different security settings.  As an
+example, a user may want to have a pod with two containers, one of which runs as root with the
+privileged setting, and one that runs as a non-root UID.  To support use cases like this, it should
+be possible to override appropriate (ie, not intrinsically pod-level) security settings for
+individual containers.
+
+## Analysis
+
+### SELinux context: pod- or container- level?
+
+Currently, SELinux context is specifiable only at the container level.  This is an inconvenient
+factoring for sharing volumes and other SELinux-secured resources between containers because there
+is no way in SELinux to share resources between processes with different MCS labels except to
+remove MCS labels from the shared resource.  This is a big security risk: _any container_ in the
+system can work with a resource which has the same SELinux context as it and no MCS labels.  Since
+we are also not interested in isolating containers in a pod from one another, the SELinux context
+should be shared by all containers in a pod to facilitate isolation from the containers in other
+pods and sharing resources amongst all the containers of a pod.
+
+### Pod-level supplemental groups
+
+The [generalized non-root and SELinux support for volumes](https://github.com/kubernetes/kubernetes/pull/12944)
+pull request describes in detail the cases for sharing volumes using supplemental groups.  The
+easiest way to share resources among containers running as different UID/primary GID combinations is
+to make them owned by a supplemental group that all containers run as.  Therefore, there should be a
+pod-level supplemental group field that cannot be overridden in containers.
+
+### Backward compatibility
+
+The existing V1 API should be backward compatible with the new API.  Backward compatibility is
+defined as:
+
+1.  Any API call (e.g. a structure POSTed to a REST endpoint) that worked before your change must
+    work the same after your change.
+2.  Any API call that uses your change must not cause problems (e.g. crash or degrade behavior) when
+    issued against servers that do not include your change.
+3.  It must be possible to round-trip your change (convert to different API versions and back) with
+    no loss of information.
+
+This essentially means that we must keep all the existing fields in place in the V1 API and make
+only additive changes.  So, in order to design the correct API changes, we must:
+
+1.  Retain the current container level `SecurityContext`
+2.  Define which fields must be controllable at the container level; this becomes the new 
```

Comment:
@smarterclayton We would need to keep the SELinux label for V1 compatibility purposes.  Maybe you're right and we can just remove that field in V2.  I'll think about it.